main.tf

```diff
@@ -40,6 +42,16 @@ resource "azurerm_storage_account" "storage_account" {
       delete_retention_policy {
         days = 365
       }
+	  dynamic cors_rule {
+		for_each = var.cors_rules
+		content {
+			allowed_headers = var.cors_rules.allowed_headers
```

Comment:
to access the value from the loop for_each you will need to access the properties like `each,value.allowed_headers` 
https://www.terraform.io/language/meta-arguments/for_each